### PR TITLE
Editor UI for custom variables

### PR DESCRIPTION
**Pull request by MartinRiese**
## Product Description

Use `uiMapList = hqImport("hqwebapp/js/ui_elements/ui-element-key-val-list")` to the custom variables UI in case lists and details. And store the variables in a dictionary instead of preformatted XML.

image

image

image

image

## Technical Summary

https://dimagi-dev.atlassian.net/browse/USH-3444

## Feature Flag

* Show editor for entering custom variables

### Automated test coverage

* Updated test that cover generating suite.xml
* Added test covering migration

### Migrations

- [x] The migrations in this code can be safely applied first independently of the code

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/app_manager/management/commands/migrate_case_list_custom_variables.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    @staticmethod
+    def parse(xml):
+        elems = etree.fromstring(f"<w>{xml}</w>")
+        return {e.tag: e.get("function") for e in elems}
+
+    @staticmethod
+    def migrate_detail_forward(detail):
+        if detail and (detail.get('custom_variables') is not None):
+            detail['custom_variables_dict'] = Command.parse(detail['custom_variables'])
+            del detail['custom_variables']
```
Comment (MartinRiese) [inline corehq/apps/app_manager/management/commands/migrate_case_list_custom_variables.py:38]: @esoergel If I understand [this guide](https://commcare-hq.readthedocs.io/migrations_in_practice.html) correctly I should remove the `del detail['custom_variables']` and do it in a separate migration later?

**Review by esoergel — COMMENTED** [review]
Nice that this can be done without messing up existing configs!

What's your rollout plan for the migration?  Try it on staging?  Run the management command prior to deploy from a private release?

File: corehq/apps/app_manager/management/commands/migrate_case_list_custom_variables.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+            return True
+        return False
+
+    @staticmethod
+    def serialize(variables_dict):
+        return "\n".join([f"<{p[0]} function=\"{p[1]}\"/>" for p in variables_dict.items()])
```
[the remark this answers is not in the document]
Comment (MartinRiese) [reply corehq/apps/app_manager/management/commands/migrate_case_list_custom_variables.py:44]: Good point.

File: corehq/apps/app_manager/views/modules.py [hunk 3]
```diff
@@ -1172,6 +1172,10 @@ def edit_module_detail_screens(request, domain, app_id, module_unique_id):
         'short': params.get("short_custom_variables", None),
         'long': params.get("long_custom_variables", None)
     }
```
[the remark this answers is not in the document]
Comment (MartinRiese) [reply corehq/apps/app_manager/views/modules.py:1174]: Thanks for the catch

File: corehq/apps/app_manager/management/commands/migrate_case_list_custom_variables.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            return True
+        return False
+
+    @staticmethod
+    def serialize(variables_dict):
+        def create_tag(name, function):
+            element = etree.Element(name, function=function)
+            return etree.tostring(element, encoding=str)
+
+        return "\n".join([create_tag(p[0], p[1]) for p in variables_dict.items()])
```
[the remark this answers is not in the document]
Comment (MartinRiese) [reply corehq/apps/app_manager/management/commands/migrate_case_list_custom_variables.py:48]: Yes please

File: corehq/apps/app_manager/models.py [hunk 5]
```diff
@@ -1951,6 +1951,7 @@ class Detail(IndexedSchema, CaseListLookupMixin):
 
     # Custom variables to add into the <variables /> node
     custom_variables = StringProperty(exclude_if_none=True)
+    custom_variables_dict = DictProperty(exclude_if_none=True)
```
[the remark this answers is not in the document]
Comment (MartinRiese) [reply corehq/apps/app_manager/models.py:1954]: Having duplicate variable names would just lead to confusion I think. And I bet formplayer would just pick the first one with a given name.

File: corehq/apps/app_manager/views/modules.py [hunk 6]
```diff
@@ -1204,8 +1204,10 @@ def edit_module_detail_screens(request, domain, app_id, module_unique_id):
     if custom_xml is not None:
         detail.short.custom_xml = custom_xml
 
-    detail.short.custom_variables_dict = custom_variables_dict['short']
-    detail.long.custom_variables_dict = custom_variables_dict['long']
+    if custom_variables_dict['short'] is not None:
+        detail.short.custom_variables_dict = custom_variables_dict['short']
```
[the remark this answers is not in the document]
Comment (MartinRiese) [reply corehq/apps/app_manager/views/modules.py:1208]: Yes. If you save the case list settings without a variable set `custom_variables_dict['short']` will be `{}` but `custom_variables_dict['long']` is `None`. So the short would get updated but not long

File: corehq/apps/app_manager/suite_xml/sections/details.py [hunk 7]
```diff
@@ -240,11 +243,10 @@ def build_detail(self, module, detail_type, detail, detail_column_infos, tabs=No
...
-        custom_variables = detail.custom_variables
-        if custom_variables:
+        custom_variables_dict = detail.custom_variables_dict
+        if custom_variables_dict:
             custom_variable_elements = [
-                variable for variable in
-                etree.fromstring("<variables>{}</variables>".format(custom_variables))
+                etree.Element(p[0], function=p[1]) for p in custom_variables_dict.items()
             ]
             d.variables.extend([
```
[the remark this answers is not in the document]
Comment (esoergel) [reply corehq/apps/app_manager/suite_xml/sections/details.py:251]: Could you comment on this?  At present this block of code does much more than it needs to - it renders a dict as a list of XML elements, which it then converts to a string, then it parses that string as `DetailVariable` XML objects.  Instead, we could just initialize the `DetailVariable`s directly from the dict.

[the remark this answers is not in the document]
Comment (MartinRiese) [reply corehq/apps/app_manager/suite_xml/sections/details.py:251]: I made the change. Not sure why it did not mark it as out of date.

[the remark this answers is not in the document]
Comment (esoergel) [reply corehq/apps/app_manager/suite_xml/sections/details.py:251]: Is it possible it was lost in a merge or something?  I don't see the change on the diff view

[the remark this answers is not in the document]
Comment (MartinRiese) [reply corehq/apps/app_manager/suite_xml/sections/details.py]: I miss-understood you before. Fixed now.

**Review by esoergel — APPROVED** [review]

**Review by esoergel — APPROVED** [review]

**Pull request merged**